core/project.go

```diff
@@ -0,0 +1,403 @@
+		return nil, fmt.Errorf("failed to unmarshal project config: %w", err)
+	}
+
+	p.Schema, err = p.getSchema(ctx, gw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get schema: %w", err)
+	}
+
+	resolvers, err := p.getResolvers(ctx, gw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get resolvers: %w", err)
+	}
+	if err := r.Add(router.StaticSchema(router.StaticSchemaParams{
+		Name:      p.Config.Name,
+		Schema:    p.Schema,
+		Resolvers: resolvers,
+	})); err != nil {
+		return nil, fmt.Errorf("failed to install project schema: %w", err)
+	}
+	return p, nil
+}
+
+func (p *Project) Commands(ctx context.Context) ([]ProjectCommand, error) {
+	schema, err := gqlparser.LoadSchema(&gqlparserast.Source{
+		Input: p.Schema,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse schema: %w", err)
+	}
+
+	commands := make([]ProjectCommand, 0, len(schema.Query.Fields))
+	for _, field := range schema.Query.Fields {
+		if field.Name == "__schema" || field.Name == "__type" {
+			continue
+		}
+		cmd, err := p.schemaToCommand(field, schema)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert schema to command: %w", err)
+		}
+		commands = append(commands, *cmd)
+	}
+
+	return commands, nil
+}
+
+// figure out if we were passed a path to a dagger.json file or a parent dir that may contain such a file
+func (p *Project) normalizeConfigPath(configPath string) string {
+	baseName := path.Base(configPath)
+	if baseName == "dagger.json" {
+		return configPath
+	}
+	return path.Join(configPath, "dagger.json")
+}
+
+func (p *Project) schemaToCommand(field *gqlparserast.FieldDefinition, schema *gqlparserast.Schema) (*ProjectCommand, error) {
+	cmd := ProjectCommand{
+		Name:        field.Name,
+		Description: field.Description,
+	}
+
+	// args become flags
+	for _, arg := range field.Arguments {
+		flag := ProjectCommandFlag{
+			Name:        arg.Name,
+			Description: arg.Description,
+		}
+		cmd.Flags = append(cmd.Flags, flag)
+	}
+
+	// subfield objects become subcommands
+	returnType := field.Type
+	if returnType.Elem != nil {
+		returnType = returnType.Elem
+	}
+	returnObj, ok := schema.Types[returnType.Name()]
+	if ok {
+		for _, subfield := range returnObj.Fields {
+			subcmd, err := p.schemaToCommand(subfield, schema)
+			if err != nil {
+				return nil, err
+			}
+			cmd.Subcommands = append(cmd.Subcommands, *subcmd)
+		}
+	}
+	return &cmd, nil
+}
+
+func (p *Project) getSchema(ctx context.Context, gw bkgw.Client) (string, error) {
+	runtimeFS, err := p.runtime(ctx, gw)
+	if err != nil {
+		return "", fmt.Errorf("failed to get runtime filesystem: %w", err)
+	}
+
+	fsState, err := runtimeFS.State()
+	if err != nil {
+		return "", fmt.Errorf("failed to get runtime filesystem state: %w", err)
+	}
+
+	projectDirState, err := p.Directory.State()
+	if err != nil {
+		return "", fmt.Errorf("failed to get project dir state: %w", err)
+	}
+
+	st := fsState.Run(
+		llb.Args([]string{entrypointPath, "-schema"}),
```

Comment:
What code does this run? Can't find where the `-schema` flag is handled. (Oh, maybe this is the magic code => schema flow?)

Reply:
Code-only schema 🤘

In Python it's handled here:

https://github.com/dagger/dagger/blob/cdd60abc190ab45ab09c858981005b77cb3e46ad/sdk/python/src/dagger/server/cli.py

Reply:
Yep, and for go it's here: https://github.com/sipsma/dagger/blob/74becfd0f7033e25fcb4e89de8d122b104445cd4/sdk/go/server.go#L77-L77

There's actually some internal docs on this interface from long ago that are still up to date: https://github.com/dagger/dagger/blob/main/core/docs/f5cij-extension_runtime_protocol.md